From 084ef12eb08c976b3c7439a6c23ccf78379e79dd Mon Sep 17 00:00:00 2001
From: Vinicius Costa Gomes <vinicius.gomes@intel.com>
Date: Wed, 23 Jun 2021 23:03:48 -0700
Subject: [PATCH 113/122] igc: Fix the case when the launchtime is equal to
 base-time

When the launchtime crosses the next "base-time" barrier, we need to
take that into account.

Signed-off-by: Vinicius Costa Gomes <vinicius.gomes@intel.com>
Signed-off-by: Muhammad Husaini Zulkifli <muhammad.husaini.zulkifli@intel.com>
---
 drivers/net/ethernet/intel/igc/igc_main.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/ethernet/intel/igc/igc_main.c b/drivers/net/ethernet/intel/igc/igc_main.c
index 2954952575d8..4a1273ff8b94 100644
--- a/drivers/net/ethernet/intel/igc/igc_main.c
+++ b/drivers/net/ethernet/intel/igc/igc_main.c
@@ -1003,6 +1003,12 @@ static __le32 igc_tx_launchtime(struct igc_adapter *adapter, ktime_t txtime)
 	 * so they point to a time one cycle in the future
 	 */
 	baset_est = ktime_add_ns(adapter->base_time, adapter->cycle_time * (n + 1));
+
+	if (baset_est == txtime) {
+		txtime -= 1000;
+		baset_est -= adapter->cycle_time;
+	}
+
 	baset_est = ktime_sub_ns(txtime, baset_est);
 	launchtime = ktime_to_timespec64(baset_est);
 
-- 
2.17.1

